pkg/infra/usagestats/service/usage_stats.go

```diff
@@ -66,47 +73,75 @@ func (uss *UsageStats) gatherMetrics(ctx context.Context, metrics map[string]int
 	metrics["stats.usagestats.debug.collect.error.count"] = errC
 }
 
+func (uss *UsageStats) runMetricsFunc(ctx context.Context, fn usagestats.MetricsFunc) (map[string]interface{}, error) {
+	start := time.Now()
+	ctx, span := uss.tracer.Start(ctx, "UsageStats.Gather")
+	fnName := runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
+	span.SetAttributes("usageStats.function", fnName, attribute.Key("usageStats.function").String(fnName))
+	defer span.End()
+
+	fnMetrics, err := fn(ctx)
+	if err != nil {
+		uss.log.FromContext(ctx).Error("Failed to fetch usage stats from provider", "error", err, "duration", time.Since(start), "function", fnName)
+		span.SetStatus(codes.Error, fmt.Sprintf("failed to fetch usage stats from provider: %v", err))
+		return nil, err
+	}
+
+	uss.log.FromContext(ctx).Debug("Successfully fetched usage stats from provider", "duration", time.Since(start), "function", fnName)
+	return fnMetrics, nil
+}
+
 func (uss *UsageStats) RegisterMetricsFunc(fn usagestats.MetricsFunc) {
 	uss.externalMetrics = append(uss.externalMetrics, fn)
 }
 
-func (uss *UsageStats) sendUsageStats(ctx context.Context) error {
+func (uss *UsageStats) sendUsageStats(ctx context.Context) (string, error) {
 	if !uss.Cfg.ReportingEnabled {
-		return nil
+		return "", nil
 	}
-
-	uss.log.Debug("Sending anonymous usage stats", "url", usageStatsURL)
+	ctx, span := uss.tracer.Start(ctx, "UsageStats.BackgroundJob")
+	defer span.End()
+	traceID := tracing.TraceIDFromContext(ctx, false)
+	uss.log.FromContext(ctx).Debug("Sending anonymous usage stats", "url", usageStatsURL)
+	start := time.Now()
 
 	report, err := uss.GetUsageReport(ctx)
 	if err != nil {
-		return err
+		return traceID, err
 	}
 
 	out, err := json.MarshalIndent(report, "", " ")
 	if err != nil {
-		return err
+		return traceID, err
 	}
 
 	data := bytes.NewBuffer(out)
-	sendUsageStats(uss, data)
-	return nil
+	err = sendUsageStats(uss, ctx, data)
+	if err != nil {
+		return traceID, err
+	}
+
+	uss.log.FromContext(ctx).Info("Sent usage stats", "duration", time.Since(start))
+	return traceID, nil
 }
 
 // sendUsageStats sends usage statistics.
 //
 // Stubbable by tests.
-var sendUsageStats = func(uss *UsageStats, data *bytes.Buffer) {
-	go func() {
-		client := http.Client{Timeout: 5 * time.Second}
-		resp, err := client.Post(usageStatsURL, "application/json", data)
-		if err != nil {
-			uss.log.Error("Failed to send usage stats", "err", err)
-			return
-		}
-		if err := resp.Body.Close(); err != nil {
-			uss.log.Warn("Failed to close response body", "err", err)
-		}
-	}()
```

Comment:
This is a notable change to get the entirety of the function inside of the span.

From what I could tell, the only reason to create a Go routine here is for a test, which could equally well be solved by creating that Go routine in the test.

Reply:
> Or maybe, if for whatever reason a panic would happen it could crash the goroutine of the background job.

Good thinking, ran that scenario through my head and I don't think that behavior would change prior/after, as a panic would escalate all the way out and crash Grafana even if it happens in a Go routine? [Spec on handling panics](https://go.dev/ref/spec#Handling_panics) says:

> While executing a function F, an explicit call to panic or a [run-time panic](https://go.dev/ref/spec#Run_time_panics) terminates the execution of F. Any functions [deferred](https://go.dev/ref/spec#Defer_statements) by F are then executed as usual. Next, any deferred functions run by F's caller are run, and so on up to any deferred by the top-level function in the executing goroutine. At that point, the program is terminated and the error condition is reported, including the value of the argument to panic.